### ISSUE TITLE
Cluster pipeline panic

### ISSUE THREAD

**Issue by dim** [go-redis/redis#86]
After plugging the new master into a real-world app, I get these:

```
2015/04/13 12:34:20 http: panic serving 127.0.0.1:47736: runtime error: index out of range
goroutine 7 [running]:
net/http.func·011()
        $GOROOT/src/net/http/server.go:1130 +0xbb
gopkg.in/redis%2ev2.(*ClusterClient).slotAddrs(0xc208096000, 0x13e8, 0x0, 0x0, 0x0)
        $GOPATH/src/gopkg.in/redis.v2/cluster.go:78 +0xee
gopkg.in/redis%2ev2.(*ClusterPipeline).Exec(0xc2086153e0, 0xc208746000, 0x2710, 0x3000, 0x0, 0x0)
        $GOPATH/src/gopkg.in/redis.v2/cluster_pipeline.go:56 +0x2b8
```

Will debug too.

**Comment by dim** [comment]
After a restart, I am now getting tons of: `redis: you open connections too fast (last error: nil)`.

**Comment by vmihailenco** [comment]
As a quick fix we should put `reloadIfDue` back to the `NewClusterClient` so `slots` are initialized.

**Comment by dim** [comment]
I'm not really sure what is going on, trying to trace it down. None of the issues were there before :(

**Comment by vmihailenco** [comment]
1. first request comes and sees that `_reload = 1`
2. reloadIfDue is called and sets `_reload = 0`
3. second request comes and sees that `_reload = 0` so it tries to access `slots`. But `reloadIfDue` has not finished its work yet, so `slots` is empty

**Comment by dim** [comment]
Sounds like it's best solved with a mutex, but that doesn't solve the `you open connections too fast` error. Is it right to call `freeConn` in the pipeline's Exec?

**Comment by dim** [comment]
Or, you could just call `setSlots(nil)` in the constructor.

**Comment by vmihailenco** [comment]
Your original version had `reloadIfDue` call in constructor. So `ClusterClient` never had uninitialized `slots`. I think that is a good quick fix.

`setSlots(nil)` will not help, because it resets slots.

**Comment by dim** [comment]
`setSlots(nil)` initializes c.slots so we will never end up with an `index out of range` but `reloadIfDue` sounds good to me too. Either way, what about `you open connections too fast` issue?

**Comment by dim** [comment]
I am not too keen on the current pipeline solution. While it is making certain things more elegant, it makes processing more verbose. Before, we abstracted all the connection/execution details into individual pipelines. Do you want me to submit another PR with the previous/working solution?

**Comment by vmihailenco** [comment]
No, fix is trivial. `freeConn` should not remove connection from the pool when `err == nil`.

```
func (c *baseClient) freeConn(cn *conn, ei error) error {
	if cn.rd.Buffered() > 0 {
		return c.connPool.Remove(cn)
	}
        if er == nil {
		return c.connPool.Put(cn)
        }
	if _, ok := ei.(redisError); ok {
		return c.connPool.Put(cn)
	}
	return c.connPool.Remove(cn)
}
```

Can you submit PR? I am not at my PC atm.

**Comment by dim** [comment]
Np

**Comment by dim** [comment]
Fixed in #87
